dali/kernels/common/split_shape.h

```diff
@@ -59,6 +63,55 @@ void split_shape(SplitFactor& split_factor, const Shape& in_shape, int min_nbloc
   }
 }
 
+/**
+ * @brief returns the dimension index with a split factor > 1
+ */
+template <typename SplitFactor>
+int LastSplitDim(const SplitFactor& split_factor) {
+  int last_split_dim = -1;
+  int ndim = dali::size(split_factor);
+  for (int d = ndim - 1; d >= 0; d--) {
+    if (split_factor[d] > 1) {
+      last_split_dim = d;
+      break;
+    }
+  }
+  return last_split_dim;
+}
+
+/**
+ * @brief Schedule work on a per-block basis
+ */
+template <int Dims, typename SplitFactor, typename ScheduleBlockFunc>
+void ScheduleBlocks(TensorShape<Dims> &start, TensorShape<Dims> &end,
+                    const SplitFactor& split_factor, int d, int max_split_dim,
+                    ScheduleBlockFunc &&func) {
+  if (d > max_split_dim || d == Dims) {
+    std::cout << "Execute \n";
+    func(start, end);
+    return;
+  }
+
+  if (split_factor[d] == 1) {
+    ScheduleBlocks(start, end, split_factor, d + 1, max_split_dim,
+                   std::forward<ScheduleBlockFunc>(func));
+    return;
+  }
+
+  int64_t start_d  = start[d];
+  int64_t extent_d = end[d] - start_d;
+  int nblocks_d = split_factor[d];
+  for (int b = 0; b < nblocks_d; b++) {
+    start[d] = extent_d *  b      / nblocks_d + start_d;
+    end[d]   = extent_d * (b + 1) / nblocks_d + start_d;
```

Comment:
All boundaries except the first and last are calculated twice. Tiny optimization:
```suggestion
  int nblocks_d = split_factor[d];
  int64_t prev_end = start_d;
  for (int b = 0; b < nblocks_d; b++) {
    start[d] = prev_end;
    end[d]   = prev_end = extent_d * (b + 1) / nblocks_d + start_d;
```